### PR TITLE
Curate titlecase

### PR DESCRIPTION
**Pull request by j23414**
### Description of proposed changes

Adds a new sub-command `augur curate titlecase` based on the [transform-string-fields script](https://github.com/nextstrain/monkeypox/blob/master/ingest/bin/transform-string-fields) in the monkeypox repo. The `augur curate normalize` sub-command has already been added based on the same script (https://github.com/nextstrain/augur/pull/1039).

See commits for details.

### Related issue(s)

Closes https://github.com/nextstrain/augur/issues/999
Related to https://github.com/nextstrain/augur/issues/860

### Testing

Added Cram tests for the new sub-command and includes doctests.

### Checklist

* [x] Add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) about the new `augur curate titlecase` sub-command.

### CODE REVIEW

File: augur/curate/titlecase.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+"""
+Standardizes string fields of the NDJSON record from stdin and outputs the
+modified record to stdout.
+"""
+import argparse
+
+import re
+import unicodedata
```
Comment (tsibley) [inline augur/curate/titlecase.py:9]: Unused import. We should have linting for this, but I guess we don't.

File: augur/curate/titlecase.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    
+    required = parser.add_argument_group(title="REQUIRED")
+    required.add_argument("--titlecase-fields", nargs="*",
+        help="List of fields to convert to titlecase.", required=True)
+
+    optional = parser.add_argument_group(title="OPTIONAL")
+    optional.add_argument("--articles", nargs="*",
+        help="List of articles that should not be cast to titlecase.")
+    optional.add_argument("--abbreviations", nargs="*",
+        help="List of abbreviations that should not be cast to titlecase, keeps uppercase.")
```
Comment (tsibley) [inline augur/curate/titlecase.py:29]: s/cast/converted/ to be consistent with the other descriptions.

File: augur/curate/titlecase.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    articles = set()
+    if args.articles:
+        articles = set(args.articles)
+
+    abbreviations = set()
+    if args.abbreviations:
+        abbreviations = set(args.abbreviations)
+
+    for index, record in enumerate(records):
+        record = record.copy()
```
Comment (tsibley) [inline augur/curate/titlecase.py:79]: Why make this copy?

File: augur/curate/titlecase.py [hunk 4]
```diff
@@ -43,21 +43,6 @@ def titlecase(text: Union[str, None], articles: Set[str] = {}, abbreviations: Se
...
-    'The Night of the Living Dead'
-
-    >>> titlecase("BRAINE-LE-COMTE, FRANCE", articles)
-    'Braine-le-Comte, France'
-
-    >>> titlecase("auvergne-RHÔNE-alpes", articles)
-    'Auvergne-Rhône-Alpes'
-
-    >>> titlecase("washington DC, usa", articles, abbreviations)
-    'Washington DC, USA'
```
Comment (tsibley) [inline augur/curate/titlecase.py]: > Migrate tests from the function comments

Why _move_ these doctests into Cram? I think they are very appropriate as unit tests of this function… (and they also nicely document its expected behaviour).

I can see the value in having similar integration tests in Cram but would think to copy or parallel these rather than move them.

File: augur/curate/titlecase.py [hunk 5]
```diff
@@ -105,8 +90,7 @@ def run(args, records):
                 if failure_reporting is DataErrorMethod.WARN:
                     print_err(f"WARNING: {failure_message}")
     
-                    # Keep track of failures for final summary
-                    failures.append((record_id, field, date_string))
+                    failures.append((record_id, field, record.get(field, "")))
```
Comment (tsibley) [inline augur/curate/titlecase.py]: This copy/paste bug fix should be part of the previous commit rather than an unrelated fixup here.

File: tests/functional/curate/cram/titlecase.t [hunk 6]
```diff
@@ -3,16 +3,44 @@ Setup
...
-  > {"record": 1, "authors": "john smith", "author2": "Jane Doe"}
-  > ~~
+Test output with lowercase articles.
 
+  $ echo '{"record": 2, "title": "the night OF THE LIVING DEAD", "location": "BRAINE-LE-COMTE, FRANCE"}' \
+  >   | ${AUGUR} curate titlecase --titlecase-fields "title" "location" --articles "a" "and" "of" "the" "le"
+  {"record": 2, "title": "The Night of the Living Dead", "location": "Braine-le-Comte, France"}
 
-Test output with Unicode normalization form "NFKC".
+Test output with abbreviations
```
Comment (tsibley) [inline tests/functional/curate/cram/titlecase.t]: Ditto with this copy/paste typo fix.

File: docs/usage/cli/curate/titlecase.rst [hunk 7]
```diff
@@ -0,0 +1,9 @@
+=================
+normalize-strings
```
Comment (tsibley) [inline docs/usage/cli/curate/titlecase.rst:2]: Title is copy/paste typo.

File: augur/curate/titlecase.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
```
Comment (tsibley) [inline augur/curate/titlecase.py:1]: No need for a shebang at the top of this file since it's not directly executable.

**Review by tsibley — COMMENTED** [review]
+1 for learning the ropes of the Augur curate framework with this `titlecase` command!

The four commits here feel like they should be just one. Most of the commits after the first are fixes for the first. We also don't typically use the `feat:`, `fix:`, etc. prefixes in Augur. The loose convention is to prefix with the major or primary "area" that the commit touches; for example `curate:`, `docs:`, `io/metadata:`, etc. Jover's commits introducing the Augur curate commands are a good recent example.

I left a bunch of notes as I read thru the PR below (but haven't run the code myself).

File: augur/curate/titlecase.py (hunk 3 above)
Comment (joverlee521) [reply augur/curate/titlecase.py:79]: Ah, this was from me in our pair-coding session.
I wanted to make sure we modified a copy of the records rather than modifying records in place.

File: augur/curate/titlecase.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        abbreviations = set(args.abbreviations)
+
+    for index, record in enumerate(records):
+        record = record.copy()
+        record_id = index
+
+        for field in args.titlecase_fields:
+            titlecased_string = titlecase(record.get(field, ""), articles, abbreviations)
+        
+            failure_message = f"Failed to titlecase {field} in record {record_id}"
```
Comment (joverlee521) [inline augur/curate/titlecase.py:85]: In error messages, we've usually used `repr()` formatted values for data specific values.
```suggestion
            failure_message = f"Failed to titlecase {field!r} in record {record_id!r}"
```

File: augur/curate/titlecase.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    
+                    failures.append((record_id, field, record.get(field, "")))
+            else:
+                record[field] = titlecased_string
+    
+        yield record
+
+    if failure_reporting is not DataErrorMethod.SILENT and failures:
+        failure_message = (
+            "Unable to change to titlecase for the following (record, field string):\n" + \
```
Comment (joverlee521) [inline augur/curate/titlecase.py:101]:
```suggestion
            "Unable to change to titlecase for the following (record, field, field value):\n" + \
```

File: augur/curate/titlecase.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        
+            failure_message = f"Failed to titlecase {field} in record {record_id}"
+            if titlecased_string is None:
+                if failure_reporting is DataErrorMethod.ERROR_FIRST:
+                    raise AugurError(failure_message)
+    
+                if failure_reporting is DataErrorMethod.WARN:
+                    print_err(f"WARNING: {failure_message}")
+    
+                    failures.append((record_id, field, record.get(field, "")))
```
Comment (joverlee521) [inline augur/curate/titlecase.py:93]: Ditto Tom's [comment](https://github.com/nextstrain/augur/pull/1197/files#r1162059998) on falling back to an empty string here instead of None. 

I think we want to report the exact value we are seeing in the data in our error messages.

File: tests/functional/curate/cram/titlecase.t [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  $ echo '{"record": 4, "location": "Auvergne-Rhône-Alpes" }' \
+  >   | ${AUGUR} curate titlecase --titlecase-fields "location"
+  {"record": 4, "location": "Auvergne-Rh\u00f4ne-Alpes"}
+
+Test output with numbers
+
+  $ echo '{"record": 5, "title": "2021 SARS-CoV"}' \
+  >   | ${AUGUR} curate titlecase --titlecase-fields "title" --abbreviations "SARS"
+  {"record": 5, "title": "2021 SARS-Cov"}
+
```
Comment (joverlee521) [inline tests/functional/curate/cram/titlecase.t:35]: _non-blocking_

I might also add a test here for trying to titlecase a field that is just a numeric value (not part of a string). 
With the default `--failure-reporting`, the numeric value field would raise an error.

**Review by joverlee521 — COMMENTED** [review]
Thank you for working through this @j23414! I hope this was a good small intro for the Augur ecosystem 🙏 

I left a couple of minor comments, but the one thing that stands out (thanks to Tom's comment on the None vs empty string values) is that the only reason for failure for this command seems to be when the field value is _not_ a string. With that in mind, I think the help/error messages can be more explicit on requiring titlecase fields to be string values. 

I think this also leads to an open question of if the command should handle null/None values separately. I imagine there's use cases for both allowing null values and for erroring on null values in the titlecase fields separate from other type errors? (This doesn't need to be addressed in this PR, just thinking out loud)

File: tests/functional/curate/cram/titlecase.t [hunk 13]
```diff
@@ -3,16 +3,44 @@ Setup
...
+  $ echo '{"record": 6, "title": "the-night-of-the-living-dead"}' \
+  >   | ${AUGUR} curate titlecase --titlecase-fields "title"
+  {"record": 6, "title": "The-Night-Of-The-Living-Dead"}
+
+# Test output with apostrophes
+# 
+#   $ echo '{"record": 5, "title": "O\u0027Reilly\u0027s"}' \
+#   >   | ${AUGUR} curate titlecase --titlecase-fields "title"
+#   {"record": 5, "title": "O'Reilly's"}
```
[the remark this answers is not in the document]
Comment (j23414) [reply tests/functional/curate/cram/titlecase.t:46]: Added because I was looking for an edgecase (e.g. "Don't capitalize 's") but then the test failed returning "O'Reilly'S" and I hadn't committed to dig into it further. 

I'm leaning toward dropping this test case.

File: augur/curate/titlecase.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    abbreviations = set()
+    if args.abbreviations:
+        abbreviations = set(args.abbreviations)
+
+    for index, record in enumerate(records):
+        record = record.copy()
+        record_id = index
+
+        for field in args.titlecase_fields:
+            titlecased_string = titlecase(record.get(field, ""), articles, abbreviations)
```
[the remark this answers is not in the document]
Comment (j23414) [reply augur/curate/titlecase.py:95]: > This would end up converting from a null value in the JSON record to an empty string value.

> change null values to empty strings

Sorry it took me a while to circle back on this. Looking at [python's dictionary get function](https://docs.python.org/3/library/stdtypes.html#dict.get), the `record.get(field,"")` is checking if a field (e.g. "division") exists but not the field's value (null or otherwise). 

Because without changing anything, titlecase errors out as expected on null and bare integer values:

```
Test case that fails on null

  $ echo '{"empty": "", "null_entry":null  }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "empty" "null_entry"
  ERROR: Failed to titlecase 'null_entry':None in record 0
  [2]

Test case that fails on a non-string int

  $ echo '{"bare_int": 2021}' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "bare_int"
  ERROR: Failed to titlecase 'bare_int':2021 in record 0
  [2]
```

The `record.get(field,"")` allows non-existent fields to pass (e.g. "division", "not exist")

```
Test cases when fields do not exist
Decide if this should error out and might affect configs like:
# https://github.com/nextstrain/monkeypox/blob/9e6c71d039cf4ef30ccb75749b634ffc9f2cf8ac/ingest/config/config.yaml#L31

  $ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
  {"region": "Europe", "country": "France", "division": "", "location": "", "not exist": ""}
```

If the line is changed to `record.get(field)`, then non-existent fields will error out.

```
$ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
ERROR: Failed to titlecase 'division':None in record 0
```

I think we want the non-existent fields to error out? This decision may affect how we call `augur curate titlecase` in [monkeypox and other pathogen ingest pipelines](https://github.com/nextstrain/monkeypox/blob/9e6c71d039cf4ef30ccb75749b634ffc9f2cf8ac/ingest/workflow/snakemake_rules/transform.smk#L78-L81), or [maybe not](https://github.com/nextstrain/monkeypox/blob/9e6c71d039cf4ef30ccb75749b634ffc9f2cf8ac/ingest/bin/transform-genbank-location#L13-L28). Open to thoughts and suggestions here.

File: augur/curate/titlecase.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+"""
+Standardizes string fields of the NDJSON record from stdin and outputs the
+modified record to stdout.
```
[the remark this answers is not in the document]
Comment (j23414) [reply augur/curate/titlecase.py]: Good catch as that description was because `normalize` and `titlecase` were combined in one script. 

I simplified the description but open to further edits. 
https://github.com/nextstrain/augur/blob/6852345c1976fda35b21583cc256a058bd5f6966/augur/curate/titlecase.py#L2

File: augur/curate/titlecase.py (hunk 1 above)
Comment (j23414) [reply augur/curate/titlecase.py]: Dropped unused import, and fixup'd in https://github.com/nextstrain/augur/pull/1197/commits/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: augur/curate/titlecase.py (hunk 2 above)
Comment (j23414) [reply augur/curate/titlecase.py]: Thanks, fixup in https://github.com/nextstrain/augur/pull/1197/commits/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: augur/curate/titlecase.py (hunk 3 above)
Comment (j23414) [reply augur/curate/titlecase.py:91]: Thanks, I left the "record.copy()" in.

File: augur/curate/titlecase.py (hunk 4 above)
Comment (j23414) [reply augur/curate/titlecase.py]: Agree, added tests back in https://github.com/nextstrain/augur/pull/1197/commits/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: augur/curate/titlecase.py (hunk 5 above)
Comment (j23414) [reply augur/curate/titlecase.py]: Fixed up, thanks! I am starting to enjoy `fixup` and `rebase`

File: tests/functional/curate/cram/titlecase.t (hunk 6 above)
Comment (j23414) [reply tests/functional/curate/cram/titlecase.t]: fixed in https://github.com/nextstrain/augur/commit/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: docs/usage/cli/curate/titlecase.rst (hunk 7 above)
Comment (j23414) [reply docs/usage/cli/curate/titlecase.rst]: fixed in https://github.com/nextstrain/augur/commit/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: tests/functional/curate/cram/titlecase.t (hunk 13 above)
[the remark this answers is not in the document]
Comment (j23414) [reply tests/functional/curate/cram/titlecase.t]: Dropped

File: augur/curate/titlecase.py (hunk 8 above)
Comment (j23414) [reply augur/curate/titlecase.py]: fixed in https://github.com/nextstrain/augur/commit/c00156ba7d31026aa89fc5ed68ae5215c09b99bd

File: augur/curate/titlecase.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        help="List of fields to convert to titlecase.", required=True)
+
+    optional = parser.add_argument_group(title="OPTIONAL")
+    optional.add_argument("--articles", nargs="*",
+        help="List of articles that should not be cast to titlecase.")
+    optional.add_argument("--abbreviations", nargs="*",
+        help="List of abbreviations that should not be cast to titlecase, keeps uppercase.")
+
+    optional.add_argument("--failure-reporting",
+        type=DataErrorMethod,
```
[the remark this answers is not in the document]
Comment (j23414) [reply augur/curate/titlecase.py]: I hope I added this correctly. Fixed here but let me know if I'm misunderstanding something here.

https://github.com/nextstrain/augur/blob/6852345c1976fda35b21583cc256a058bd5f6966/augur/curate/titlecase.py#L28-L33

File: tests/functional/curate/cram/titlecase.t (hunk 12 above)
Comment (j23414) [reply tests/functional/curate/cram/titlecase.t:36]: Neat, added as part of https://github.com/nextstrain/augur/pull/1197/commits/3c20dcb964ddfa5d40047eff92e223aab4a2a766

**Comment by j23414** [comment]
Thanks for the reviews @tsibley and @joverlee521 ! I think I addressed most edits but let me know if I've misunderstood something. 

The only remaining thing is handling `record.get(field,"")` as referenced in the following threads:

* https://github.com/nextstrain/augur/pull/1197#discussion_r1162059998
* https://github.com/nextstrain/augur/pull/1197#discussion_r1163327475
* https://github.com/nextstrain/augur/pull/1197#discussion_r1163346828

My main understanding of the ramifications of `record.get(field,"")` is in:

* https://github.com/nextstrain/augur/pull/1197#discussion_r1174194427

Do you think `record.get(field,"")` should be changed to `record.get(field)` so that non-existent fields error out?

```
$ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
ERROR: Failed to titlecase 'division':None in record 0
```

If so, happy to add it as another commit to this PR. Otherwise this PR is ready for review.

File: augur/curate/titlecase.py (hunk 3 above)
Comment (tsibley) [reply augur/curate/titlecase.py:91]: Nod. We may want to reconsider making a copy in the future, as copies tend to be expensive.

**Comment by tsibley** [comment]
I'd think to make non-existent and existent-but-`null`-valued fields pass thru unchanged by default. And then, as Jover mentions, optionally (and not necessarily as part of this PR), allow those cases to instead be reported as warnings or made into fatal errors.

I'd also think that, by default, fields that exist with a non-null value but one that's not a string should produce an error.

File: augur/curate/titlecase.py [hunk 17]
```diff
@@ -93,7 +93,7 @@ def run(args, records):
 
         for field in args.titlecase_fields:
             titlecased_string = titlecase(record.get(field, ""), articles, abbreviations)
-            field_value = record.get(field, "")
+            field_value = record.get(field, "Not a field")
```
Comment (tsibley) [inline augur/curate/titlecase.py:96]: Using strings like this `"Not a field"` as placeholders can be confusing. For example, in the messages below which use `{field_value!r}` the placeholder value will represented as if it was the actual value found in the record. Let's fix this while addressing the missing/null-valued field behaviour.

File: augur/curate/titlecase.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/curate/titlecase.py]: Right, using `record.get(field)` will not allow you to differentiate when a field simply does not exist or when the field value is null. We can separate these two if we want different behaviors or different error messages for each case:
```python
if field not in record:
    # do something if field does not exist
elif record[field] is None:
    # do something if field value is null
```

File: augur/curate/titlecase.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
+"""
+Applies titlecase to string fields in a metadata record
+"""
+import argparse
```
Comment (joverlee521) [inline augur/curate/titlecase.py:4]: Unused import

File: augur/curate/titlecase.py (hunk 17 above)
Comment (j23414) [reply augur/curate/titlecase.py:94]: Agree with you that the placeholder can be confusing. Just documenting my thought process that this was a hacky way to get a descriptive error message below.

```
  $ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
  ERROR: Failed to titlecase 'division':'Not a field' in record 0
  [2]
```

Thanks, I'll revisit this and change it

**Comment by j23414** [comment]
From https://github.com/nextstrain/augur/pull/1197#issuecomment-1520616275, splitting this out to a separate thread:

> I'd also think that, by default, fields that exist with a non-null value but one that's not a string should produce an error.

I've got the the following two tests which do produce errors:

```
Test case that fails on a non-string int

  $ echo '{"bare_int": 2021}' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "bare_int"
  ERROR: Failed to titlecase 'bare_int':2021 in record 0
  [2]
```

@tsibley Is there another test case you were referring to by "not a string" & "non-null value"? Happy to add a test, I'm just drawing a blank on examples.

File: augur/curate/titlecase.py (hunk 17 above)
Comment (j23414) [reply augur/curate/titlecase.py]: Clarification question about:

> I'd think to make non-existent and existent-but-null-valued fields pass thru unchanged by default.

The original script ([transform-string-fields](https://github.com/nextstrain/monkeypox/blob/master/ingest/bin/transform-string-fields)) allows non-existent fields to get passed through and added to the final entry.

```
  $ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
  {"region": "Europe", "country": "France", "division": "", "location": "", "not exist": ""}
```

Just making sure this is what you mean by "non-existent pass thru unchanged by default". Or were you expecting:

```
  $ echo '{"region":"europe", "country":"france" }' \
  >   | ${AUGUR} curate titlecase --titlecase-fields "region" "country" "division" "location" "not exist"
  {"region": "Europe", "country": "France"}
```

I'm adding these as test cases to make sure we have documented edgecases.

Comment (tsibley) [reply augur/curate/titlecase.py]: I'd expect the latter, i.e. fields which didn't exist in an input record are _not_ added to output records. Something like:

```
  $ echo '{"a": "hello world", "b": null}' \
  >   | ${AUGUR} curate titlecase --titlecase-fields a b c
  {"a": "Hello World", "b": null}
```

**Comment by tsibley** [comment]
> @tsibley Is there another test case you were referring to by "not a string" & "non-null value"? Happy to add a test, I'm just drawing a blank on examples.

Nope! That's the sort of thing I meant. I was describing the behaviour I expected, not how it differed from the current behaviour in the PR.

I do think that the error message like:

    Failed to titlecase 'bare_int':2021 in record 0

would be improved by explaining _why_ titlecasing failed.

**Comment by j23414** [comment]
Thanks @joverlee521  and @tsibley! Added the remaining changes with @joverlee521.

* [fields which didn't exist in an input record are not added to output records](https://github.com/nextstrain/augur/pull/1197#discussion_r1177066060)

https://github.com/nextstrain/augur/blob/4a39f5b470afd81fb791fc82b20283c08432d369/tests/functional/curate/cram/titlecase.t#L64-L68

* [existent-but-null-valued fields pass thru unchanged by default](https://github.com/nextstrain/augur/pull/1197#issuecomment-1520616275)

https://github.com/nextstrain/augur/blob/4a39f5b470afd81fb791fc82b20283c08432d369/tests/functional/curate/cram/titlecase.t#L43-L47

* [Error messages that explain why bare_int (or other non-string values) fail](https://github.com/nextstrain/augur/pull/1197#issuecomment-1522431548)

https://github.com/nextstrain/augur/blob/4a39f5b470afd81fb791fc82b20283c08432d369/tests/functional/curate/cram/titlecase.t#L50-L62

This PR is ready for review

File: augur/curate/titlecase.py [hunk 19]
```diff
@@ -98,7 +98,7 @@ def run(args, records):
 
             titlecased_string = titlecase(record[field], articles, abbreviations)
         
-            failure_message = f"Failed to titlecase {field!r}:{record.get(field)!r} in record {record_id!r}"
+            failure_message = f"Failed to titlecase {field!r}:{record.get(field)!r} in record {record_id!r} because the value is a {type(record.get(field))!r} and is not a string."
```
Comment (joverlee521) [inline augur/curate/titlecase.py:101]: Using `type(...).__name__` will print the class name without the `<class ... >` bit in the message.

Comment (j23414) [reply augur/curate/titlecase.py:101]: Thanks! I'll try it

Comment (j23414) [reply augur/curate/titlecase.py]: After [some reading](https://stackoverflow.com/questions/510972/getting-the-class-name-of-an-instance), [fixed](https://github.com/nextstrain/augur/pull/1197/commits/24fe373dd003eb5671b2e1b0c19a04c1185c06c7) and tested. Thanks!

File: tests/functional/curate/cram/titlecase.t
Comment (joverlee521) [inline tests/functional/curate/cram/titlecase.t:1]: Oh, I just realized that we are not testing the `--failure-reporting` option at all in the Cram tests. 
It might be worth it to add at least one test per option (similar to the [format-dates tests](https://github.com/nextstrain/augur/blob/9ef471129fc048f596952af5cbee4cd87feac20b/tests/functional/curate/cram/format_dates.t#L29-L65)) to ensure the correct behavior.

File: augur/curate/titlecase.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+            failure_message = f"Failed to titlecase {field!r}:{record.get(field)!r} in record {record_id!r} because the value is a {type(record.get(field)).__name__!r} and is not a string."
+            if titlecased_string is None:
+                if failure_reporting is DataErrorMethod.ERROR_FIRST:
+                    raise AugurError(failure_message)
+    
+                if failure_reporting is DataErrorMethod.WARN:
+                    print_err(f"WARNING: {failure_message}")
+    
+                    # Keep track of failures for final summary
+                    failures.append((record_id, field, record.get(field)))
```
Comment (joverlee521) [inline augur/curate/titlecase.py:110]: Just realized that the `error_all` failure method is not working as expected because the failures only get appended if the failure method is `warn`. 

Moving the append out of the if statement here would fix this:
```suggestion
                if failure_reporting is DataErrorMethod.WARN:
                    print_err(f"WARNING: {failure_message}")

                # Keep track of failures for final summary
                failures.append((record_id, field, record.get(field)))
```

**Review by joverlee521 — CHANGES_REQUESTED** [review]
Thanks for continue to push on this @j23414 🙏 

There's a final bug that we missed that needs to be fixed and tested in the Cram tests, but everything else looks good to me.

File: tests/functional/curate/cram/titlecase.t
Comment (j23414) [reply tests/functional/curate/cram/titlecase.t:1]: Thanks for flagging and the discussion on `--failure-reporting`. Incorporated changes in https://github.com/nextstrain/augur/pull/1197/commits/9b550d77b275a093cd6256c4bbf333ab4ba719fd

Comment (joverlee521) [reply tests/functional/curate/cram/titlecase.t:1]: Thanks for adding the tests. The changes look good to me. 

---
I think the only things left are to fixup the commits and add an entry to the Changelog.

Comment (j23414) [reply tests/functional/curate/cram/titlecase.t:1]: Fixup'd and added changelog with https://github.com/nextstrain/augur/pull/1197/commits/5896d02aeabee97c1c3d555254ace5b0cc7cd66d but feel free to suggest different wording.

**Review by joverlee521 — APPROVED** [review]
Thanks @j23414! This just requires a final rebase to fix the changelog merge conflicts and should be good to merge.
